src/plugins/discover/public/application/angular/discover.js

```diff
@@ -125,24 +126,7 @@ app.config(($routeProvider) => {
           const { appStateContainer } = getState({ history });
           const { index } = appStateContainer.getState();
           return Promise.props({
-            ip: indexPatterns.getCache().then((indexPatternList) => {
-              /**
-               *  In making the indexPattern modifiable it was placed in appState. Unfortunately,
-               *  the load order of AppState conflicts with the load order of many other things
-               *  so in order to get the name of the index we should use, and to switch to the
-               *  default if necessary, we parse the appState with a temporary State object and
-               *  then destroy it immediatly after we're done
-               *
-               *  @type {State}
-               */
-              const id = getIndexPatternId(index, indexPatternList, config.get('defaultIndex'));
-              return Promise.props({
-                list: indexPatternList,
-                loaded: indexPatterns.get(id),
-                stateVal: index,
-                stateValFound: !!index && id === index,
-              });
-            }),
+            ip: loadIndexPattern(index, data.indexPatterns, config),
```

Comment:
Note that the refactoring here is a first step, separate loading and resolving shouldn't be necessary, it improving it step by step